db/hints/sync_point.cc

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (C) 2021-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <exception>
+#include <unordered_set>
+
+#include <seastar/core/simple-stream.hh>
+
+#include "db/hints/sync_point.hh"
+#include "gms/inet_address_serializer.hh"
+#include "idl/uuid.dist.hh"
+#include "idl/uuid.dist.impl.hh"
+#include "idl/replay_position.dist.hh"
+#include "idl/replay_position.dist.impl.hh"
+#include "idl/hinted_handoff.idl.hh"
+#include "idl/hinted_handoff.dist.hh"
+#include "idl/hinted_handoff.dist.impl.hh"
+#include "serializer.hh"
+#include "serializer_impl.hh"
+#include "bytes.hh"
+
+namespace db {
+namespace hints {
+
+// Format V1 (encoded as a hexadecimal string):
```

Comment:
base64 will result in smaller things, we have functions for it.

Reply:
We do have functions for encoding in base64, but I was hesitant to use it because they are defined in the alternator directory. It feels strange for hints to depend on parts of the alternator, but maybe it's not harmful. I can switch to base64.

Reply:
@piodul feel free to move them out. It already happened with our rapidjson wrappers.

Reply:
Looks good, but we just need to remember in the future not to modify these functions in a way that can break Alternator. We do have a few tests in test/alternator which use "bytes" types (that use this base64 encoding) that should hopefully break if these utilities break - but I'm not sure how much of base64 corner cases these tests actually check (it certainly checks some, but possibly not all). We should have written a unit test in C++ directly for the base64 encoding and decoding functions that attempts to have full code coverage - or even better, use @denesb 's code coverage analyzer to make sure that test/alternator cover 100% of the base64 code.

Another minor problem with the base64.{hh,cc} is that its API is was designed to fit Alternator's needs exactly - in particular base64_encode returns std::string instead of seastar::sstring because this is what Altenator uses (for the reason why I want to move away from sstring, see https://github.com/scylladb/seastar/issues/634). It also has some peculiar APIs involving the subtly different "bytes" and "strings".